feature:

Sweta304 opened the pull request:
- добавлены классы для обновления, хранения и поиска объектов (storage package)
- добавлена возможность добавления в друзья
- добавлена возможность поставить фильму лайк

Sweta304 committed:
feature:
- добавлены классы для обновления, хранения и поиска объектов (storage package)
- добавлена возможность добавления в друзья
- добавлена возможность поставить фильму лайк

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,65 +1,75 @@
-    private Map<Long, Film> films = new HashMap<>();
-    private long lastId = 0;
+
+    private final FilmStorage filmStorage;
```
В контроллере все взаимодействие с сохраненными фильмами должно проходить через filmService, а не напрямую

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,65 +1,75 @@
-            log.info("добавлен новый фильм с id {}", film.getId());
-        }
-        return film;
+        return filmStorage.addFilm(film);
```
Здесь и далее, где используется filmStorage, желательно использовать вызов сервиса вместо стораджа

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,65 +1,75 @@
+        if (filmStorage.getFilmById(id) == null) {
+            throw new FilmNotFoundException("фильма не существует");
+        }
+        if (!filmStorage.getFilmById(id).getUsersLikes().contains(userId)) {
```
Будет лучше filmStorage.findById() вызывать 1 раз и сохранить результат в переменную вместо множества вызовов

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,65 +1,75 @@
+
+    @PutMapping("/{id}/like/{userId}")
+    public long addLike(@PathVariable long id, @PathVariable long userId) throws FilmNotFoundException{
+        if (filmStorage.getFilmById(id) == null) {
```
Будет лучше filmStorage.findById() вызывать 1 раз и сохранить результат в переменную вместо множества вызовов

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,65 +1,75 @@
-        return ++lastId;
+    @GetMapping("/popular")
+    public List<Film> getTopFilms(@RequestParam(required = false, defaultValue = "10") int count) {
+        return filmService.getTopFilms(count);
```
А что будет если прислать count нулевой или отрицательный?

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,68 +1,89 @@
-    //    private final List<User> users = new ArrayList<>();
-    private Map<Long, User> users = new HashMap<>();
-    private long lastId = 0;
+    private final UserStorage userStorage;
```
Здесь тот же комментарий, что и в FilmController

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,68 +1,89 @@
-        return ++lastId;
+    @PutMapping("/{id}/friends/{friendId}")
+    public User addFriend(@PathVariable long id, @PathVariable long friendId) throws UserNotFoundException {
+        if (userStorage.getUsers().get(id) == null || userStorage.getUsers().get(friendId) == null) {
```
Лучше сохранять в переменные результаты вызовов методов, чтобы не повторять вычисления/поиск

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,68 +1,89 @@
+        if (userStorage.getUsers().get(id) == null || userStorage.getUsers().get(friendId) == null) {
+            throw new UserNotFoundException("пользователя для добавления друга не существует");
+        }
+        return userService.addFriend(userStorage.getUsers().get(id), userStorage.getUsers().get(friendId));
```
Может будет достаточно передавать Id (а не объекты User), и сделать проверку в сервисе, а не контроллере?

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,68 +1,89 @@
 
+    @DeleteMapping("/{id}/friends/{friendId}")
+    public User deleteFriend(@PathVariable long id, @PathVariable long friendId) {
+        return userService.deleteFriend(userStorage.getUsers().get(id), userStorage.getUsers().get(friendId));
```
Здесь (а лучше в сервисе) наоборот проверка на существования юзеров пропущена. И тоже может будет достаточно передавать Id (а не объекты User)

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,40 @@
+        return commonFriends;
+    }
+
+    public Set<Long> getFriends (User user) {
```
Нужен ли такой метод, если это просто вызов метода на самом объекте? Выглядит лишним. Только если в качестве аргумента принимать только id, то тогда еще можно понять его присутствие

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,68 +1,89 @@
+    @GetMapping("/{id}/friends")
+    public Set<User> getFriends(@PathVariable long id) {
+        Set<Long> friendsIdList = userService.getFriends(userStorage.getUsers().get(id));
+        Set<User> friendsList = new HashSet<>();
```
friendsList, который Set немного смущает :)

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,40 @@
+        return user;
+    }
+
+    public Set<Long> getCommonFriends(User user, User friend) throws UserNotFoundException{
```
А кто может бросить UserNotFoundException? Я не вижу внутри метода ничего такого

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmStorage.java:
```diff
@@ -0,0 +1,22 @@
+
+    List<Film> getAllFilms();
+
+    Map<Long, Film> getFilms();
```
Этот метод выглядит как раскрытие способа хранения данных :) А что если потом будут другие реализации Storage, не с помощью мапы? Как тогда такой метод реализовать?

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserStorage.java:
```diff
@@ -0,0 +1,22 @@
+
+    List<User> getAllUsers();
+
+    Map<Long, User> getUsers();
```
Здесь такой же комментарий, как и с FilmsStorage - такой метод лишний в интерфейсе

alexache reviewed (changes requested):
Есть вопросы к разделению между контроллерами, сервисами и стораджем

Sweta304 committed:
refactor:
- из интерфейсов FilmStorage, UserStorage удалены лишние методы
- FilmController, UserController теперь используют только сервисы, использование storage перенесено в сервисы

Sweta304 committed:
- добавлена аннотация @Autowired

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -19,23 +23,47 @@ public FilmService(FilmStorage filmStorage) {
-    public long addLike(Film film, long userId) {
+    public long addLike(long filmId, long userId) throws FilmNotFoundException {
+        Film film = getFilmById(filmId);
         film.addLike(userId);
```
(можно лучше) В FIlmService можно было бы добавить еще userStorage и проверять существование юзера у стораджа (или сервиса)

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryUserStorage.java:
```diff
@@ -64,11 +64,11 @@ public List<User> getAllUsers() {
-    public User getUserById (long id) throws UserNotFoundException {
-        if (getUsers().get(id) == null) {
+    public User getUserById(long id) throws UserNotFoundException {
+        if (users.get(id) == null) {
```
Промежуточные результаты лучше сохранять в переменную

alexache reviewed (approved):
Замечания устранены, все ок!

Sweta304 committed:
refactor
